doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md

```diff
@@ -0,0 +1,97 @@
+
+One example in the code gen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised.  Then it will branch to an out of line, or a snippet of code that jumps into the VM to throw the exception that needs to be thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet code is put at the end of the code section, or at the end of the generated code for that method, aka out of line, so that it does not interfere with the code that we believe will actually run.
+
+Another example is block ordering and extension of basic blocks in IL. Both uses the same logic to arrange the blocks. For example there is a block ending with `if`. One path is marked as cold. Another path is not marked as cold. Block ordering takes the cold path out of the mainline, puts it at the end of the method, and keeps the non-cold path as the fall through. This way it avoids consuming branch prediction and instruction cache resources.
+
+
+# Global register dependency
+
+`GlRegDeps` at its core is really a code gen register dependency. It associates virtual registers with real registers, or pairs virtual registers with real registers.
+
+Register dependencies enforce constraints. It says that this value has to be in this register at this time. Global register dependency is fundamentally about conveying particular values in particular registers across the block boundaries. The values and registers are expressed by `GlRegDeps` from one basic block to another.
+
+Most common control flow opcodes are: `if`, `goto`, `return`, `athrow`, `switch`. They should be the last tree before the `BBEnd`. They should have their own `GlRegDeps` for the branch target if required. The `GlRegDeps` under `BBEnd` is for the fall through case.
+
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling could come with a cost.
+
+The local register allocator and the code gen use the reference count for `GlRegDeps` to determine when a register is no longer needed. It needs to be set up correctly, otherwise it could result in registers being overwritten with some other values prematurely.
+
+The `GlRegDeps` nodes should have been evaluated before they get to `GlRegDeps`. The `GlRegDeps` job is to take values that are already in virtual registers and simply map them to real registers. `GlRegDeps` is only about generating the register dependency. It should not be about generating any other instructions.
+
```

Comment:
The only other instruction that a `GlRegDeps` evaluation can result in is a register-register move since we can have a situation where the same value (same IL node) goes out in two (or more) different global registers. In such a case we cannot have the same virtual register be mapped to two (or more) real registers in the codegen register dependency since this is a condition that cannot be satisfied. By adding a new virtual register and inserting a register-register copy instruction the codegen register dependency can use different virtual/real registers.